metricsbp/sys_stats.go

```diff
@@ -0,0 +1,113 @@
+package metricsbp
+
+import (
+	"runtime"
+	"time"
+)
+
+// SysStatsTickerInterval is the interval we pull and report sys stats.
+// Default is 10 seconds.
+var SysStatsTickerInterval = time.Second * 10
+
+type cpuStats struct {
+	NumGoroutine int
+	NumCgoCall   int64
+}
+
+func pullRuntimeStats() (cpu cpuStats, mem runtime.MemStats) {
+	runtime.ReadMemStats(&mem)
+	cpu.NumGoroutine = runtime.NumGoroutine()
+	cpu.NumCgoCall = runtime.NumCgoCall()
+	return
+}
+
+// RunSysStats starts a goroutine to periodically pull and report sys stats.
+//
+// StopReporting will also stop this goroutine.
+func (st Statsd) RunSysStats() {
+	// init the gauges
+	// cpu
+	cpuGoroutines := st.Gauge("cpu.goroutines")
+	cpuCgoCalls := st.Gauge("cpu.cgo_calls")
+	// gc
+	gcSys := st.Gauge("mem.gc.sys")
+	gcNext := st.Gauge("mem.gc.next")
+	gcLast := st.Gauge("mem.gc.last")
+	gcPauseTotal := st.Gauge("mem.gc.pause_total")
+	gcPause := st.Gauge("mem.gc.pause")
+	gcCount := st.Gauge("mem.gc.count")
+	// general
+	memAlloc := st.Gauge("mem.alloc")
+	memTotal := st.Gauge("mem.total")
+	memSys := st.Gauge("mem.sys")
+	memLookups := st.Gauge("mem.lookups")
+	memMalloc := st.Gauge("mem.malloc")
+	memFrees := st.Gauge("mem.frees")
+	// heap
+	heapAlloc := st.Gauge("mem.heap.alloc")
+	heapSys := st.Gauge("mem.heap.sys")
+	heapIdle := st.Gauge("mem.heap.idle")
+	heapInuse := st.Gauge("mem.heap.inuse")
+	heapReleased := st.Gauge("mem.heap.released")
+	heapObjects := st.Gauge("mem.heap.objects")
+	// stack
+	stackInuse := st.Gauge("mem.stack.inuse")
+	stackSys := st.Gauge("mem.stack.sys")
+	mspanInuse := st.Gauge("mem.stack.mspan_inuse")
+	mspanSys := st.Gauge("mem.stack.mspan_sys")
+	mcacheInuse := st.Gauge("mem.stack.mcache_inuse")
+	mcacheSys := st.Gauge("mem.stack.mcache_sys")
+	// other
+	memOther := st.Gauge("mem.othersys")
+
+	st.wg.Add(1)
+	go func() {
+		defer st.wg.Done()
+
+		ticker := time.NewTicker(SysStatsTickerInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-st.ctx.Done():
+				return
+			case <-ticker.C:
```

Comment:
What is the behavior here if we update the ticker interval, does that get pulled in or would the goroutine need to be restarted?

Reply:
Updating the interval will have no effect here. You need to do that before calling `RunSysStats`.